### ISSUE TITLE
[Technical] Implementing User API Token

### ISSUE THREAD

**Issue by FenrirUnbound** [screwdriver-cd/screwdriver#532]
#As a one-time experiment, separate the technical work from the user-facing feature work. The objective is: separate the levels of discussion, allowing dedicated spaces for technical collaboration and feature-level conversations.

Developer updates & technical conversations would be attached here. 

# Summary
**High-level description**: Users can generate an access token that can be used to interact with the v4 API, as an alternative to OAuth (think Github access token)
**Full-stack project** (front-end for user-friendly token generation; server side; DB to store issued tokens and track valid/invalid ones); good mix of front/back
**Story size**: Medium-sized
**Technologies used**: Node.JS
**Impact/value**: Better security

Main Feature (with rules/requirements) - #524 

---
## Preparation

- [x] Feature file (#538)

### API

- [x] Tokens endpoint
   - [x] Create tokens
   - [x] List tokens
      - [x] List tokens via pagination
   - [x] Update token name/description
   - [x] Delete tokens
   - [x] Update "last seen" token property

### Data Schema

- [x] Schema for Tokens endpoint (API)
- [x] Schema for tokens model

### Models

- [x] Tokens model

### UI

- [x] Personal Preferences page
 - [ ] Re-scope after API is completed

## Outcome

- [x] Passing acceptance tests described in feature file
   - [x] (Optional) Tech Debt: Use issued token in acceptance tests
- [x] Documentation
- [x] Blog post
- [ ] Optional: Video tutorial

## References
* Github Personal Access tokens (for reference)
   * https://github.com/settings/tokens
* OAuth testing
   * CasperJS
      * https://github.com/oauth-io/testing-oauth
   * Zombie.js
      * https://blog.bandwidth.com/testing-oauth-integration-flows-with-zombie-js/
      * https://github.com/a-type/zombie-oauth-example

### 2017-04-24

* Initial spec
* Added blog post as a requirement for definition of done (tkyi)
* Added feature file criteria in addition to OAuth testing references
   * Included ZombieJS references

### 2017-04-25

* Add Summary for context
* Tokens table details
* Add update endpoint for API
* Add schema for token table?
* Add Github personal tokens link under References

### 2017-04-26

* Included UI work

### 2017-05-10
* Removed requirements for datastore, and data-schema table. We don't need it. 

### 2017-05-30

* Included pagination for listing the user's tokens

**Comment by tkyi** [comment]
Should we also add feature file to this list?

**Comment by FenrirUnbound** [comment]
> Should we also add feature file to this list?

I vaguely recall that the functional tests cannot follow the OAuth flow for authentication. Do we know how to do the OAuth flow without a human?

**Comment by petey** [comment]
Here's some example code: https://github.com/oauth-io/testing-oauth

**Comment by FenrirUnbound** [comment]
## SUMMARY 2017-04-26

* Drafted feature file - #538

**Comment by ian-fox** [comment]
So far:
* Data schema is done
* Models are being reviewed

**Comment by minz1027** [comment]
Updates 05/16/2017:

The personal access token is just a JWT token that never expires.

### Create flow
1. POST token - Generate a JWT token that never expires and store it in the database.

### Usage flow
1. Check if the token has been deleted first and then verify the JWT
    - If deleted, invalidate the token
    - Notes: there is no way to revoke the jwt token directly
Ref: https://dev.to/neilmadden/7-best-practices-for-json-web-tokens
2. The rest should be the same as the JWT token we have now

### TODO:
- [ ] Add a new JWT auth strategy with no `maxAge` and a custom validate function: https://github.com/screwdriver-cd/screwdriver/blob/master/plugins/auth/index.js#L105

- [ ] Add and expose function `generatePersonalToken` with no `expirationIn` https://github.com/screwdriver-cd/screwdriver/blob/master/plugins/auth/index.js#L74

**Comment by ian-fox** [comment]
Update 05/22/2017:

API tokens should not be JWT's. Back to storing token values (encrypted with iron) and checking against those with a login route, where the API token will be traded for a 12 hour JWT.

**Comment by ian-fox** [comment]
For authenticating, my revised understanding is that the user will make a POST to a login route with their API token, and be issued a 12 hour JWT. That POST request needs to have the token, but it may need some sort of secondary identifying information. 

This is because Iron's sealing depends on time: sealing the same string on two occasions doesn't give two copies of the same sealed string. That means we can't seal the token that comes in and query the database to find a match, every token we want to check it against has to be unsealed. So unless we want to unseal every token on the cluster every time an API tries to log in, it will need some manner of narrowing that list down for us.

This could be in the form of the ID of the token, the ID of the user (so we only have to unseal that user's tokens).

**Comment by ian-fox** [comment]
Or, alternatively, we could hash the tokens with some sort of algorithm other than iron that gives the same result each time, and then could hash incoming tokens and find the match in the database.

**Comment by ian-fox** [comment]
Update 05/24/2017:

- [ ] [PR is open](https://github.com/screwdriver-cd/models/pull/163) in the models repo for token model to store a hashed value.
- [ ] Once that is approved, code is ready for a second PR to the models repo adding the ability to get a user by token value in the user factory.
- [ ] Once that one is approved, code is ready for a PR to the screwdriver repo for authenticating with a token.
- [ ] Routes are in progress.

**Comment by ian-fox** [comment]
Are there any sort of mockups of how it's expected to look with regards to layout?

**Comment by FenrirUnbound** [comment]
> Are there any sort of mockups of how it's expected to look with regards to layout?

Currently, no. You get to learn about those 😺

**Comment by FenrirUnbound** [comment]
As of now, there's a soft cap of 50 tokens. When we visit the UI mocks, we may consider looking into how to remove that cap and display it to the user

**Comment by ian-fox** [comment]
Related to https://github.com/screwdriver-cd/screwdriver/pull/568

**Comment by ian-fox** [comment]
Update from standup today:
* Tokens are currently passed in the query string, as the auth header is used by hapi jwt
* If they are restricted to the query string, they should not use base64 as it is not URL-safe. 
* Now investigating other possible ways to have both the hapi jwt and the hapi bearer plugins check the `authorization` header so that tokens can remain base64.

**Comment by ian-fox** [comment]
Things still do to with UI:

- [ ] Use CSS to make things look pretty
- [ ] Be able to update name/description
- [ ] Redirect to the login page if there's no active session
- [ ] Make sure errors are formatted nicely
- [ ] Confirmation modal before deleting/refreshing a token
- [ ] Sort tokens by different fields
- [ ] Tests

Already done:
- [x] Add "user settings" page to menu-bar
- [x] Display tokens on user page
- [x] Operations on tokens:
  - [x] Create tokens
  - [x] Refresh tokens
  - [x] Delete tokens
